library/core/src/task/wake.rs

```diff
@@ -481,18 +481,13 @@ impl Waker {
     /// use std::future::Future;
     /// use std::task;
     ///
-    /// let mut cx = task::Context::from_waker(task::Waker::noop());
+    /// let mut cx = task::Context::from_waker(task::Waker::NOOP);
     ///
     /// let mut future = Box::pin(async { 10 });
     /// assert_eq!(future.as_mut().poll(&mut cx), task::Poll::Ready(10));
     /// ```
-    #[inline]
-    #[must_use]
     #[unstable(feature = "noop_waker", issue = "98286")]
-    pub const fn noop() -> &'static Waker {
-        const WAKER: &Waker = &Waker { waker: RawWaker::NOOP };
-        WAKER
-    }
+    pub const NOOP: &'static Waker = &Waker { waker: RawWaker::NOOP };
```

Comment:
FWIW, since constants get promoted, it'd now also be possible to change this back to  habe type `Waker` and use it as
```rust
let ctx = &mut Context::from_waker(&Waker::NOOP);
```

Reply:
[I already tried that initially and the compiler rejects it.](https://github.com/rust-lang/rust/issues/98286#issuecomment-1891124713)

> I prepared a change which replaced `Waker::noop()` with `Waker::NOOP`. Unfortunately, it seems that this doesn't enable `&NOOP` usage:
> ```rust
> let mut cx = &mut Context::from_waker(&Waker::NOOP);
> cx.waker().wake_by_ref();
> ```
> ```
> error[E0716]: temporary value dropped while borrowed
>    --> src/wake.rs:404:44
>     |
> 404 |     let mut cx = &mut Context::from_waker(&Waker::NOOP);
>     |                                            ^^^^^^^^^^^ - temporary value is freed at the end of this statement
> ```
> If I understand correctly now, this is not possible [because](https://doc.rust-lang.org/reference/destructors.html#constant-promotion) the `Waker` has a destructor. Therefore, since use via constant promotion is not possible, it seems wise to offer a `&Waker` value, in which case making it a constant rather than a `const fn` doesn't provide any advantages (but still might be a choice of API aesthetics).

Reply:
Ah, true, if it has a destructor it cannot get promoted. Good point.